### PR TITLE
Fix Bugs and Add Labels

### LINKED ISSUE #6: Removed instances turn idle, triggering terminate, which crashes
https://github.com/awslabs/ec2-spot-jenkins-plugin/blob/79773e0525bff5a864782e56606fe6fc58612a42/src/main/java/com/amazon/jenkins/ec2fleet/IdleRetentionStrategy.java#L28

Instead of crashing when it sees a dying instance, `terminateInstance` should just remove that node from Jenkins.

### PR DESCRIPTION
**Pull request by jzila**
Fixes #6

### CODE REVIEW

File: src/main/java/com/amazon/jenkins/ec2fleet/EC2FleetCloud.java [hunk 1]
```diff
@@ -244,7 +244,7 @@ private void addNewSlave(final AmazonEC2 ec2, final String instanceId) throws Ex
         //noinspection SynchronizationOnLocalVariableOrMethodParameter
         synchronized (jenkins) {
             // Try to avoid duplicate nodes
-            final Node n = jenkins.getNode(name);
```
Comment (jzila) [inline src/main/java/com/amazon/jenkins/ec2fleet/EC2FleetCloud.java:247]: not sure where this `name` variable came from

**Comment by jzila** [comment]
cc @Cyberax for review. I've been running this for a few weeks and it works well.

**Comment by eyalstoler** [comment]
@jzila :+1: I will try it out.

**Comment by jzila** [comment]
Side note: this makes it really easy to migrate to an updated cluster whenever you need to make changes to the AMI: just change the label to "deprecated" or something like that, and new jobs won't be scheduled on it. Then introduce the new cluster on the new AMI with the correct label.

File: src/main/java/com/amazon/jenkins/ec2fleet/EC2FleetCloud.java [hunk 2]
```diff
@@ -233,18 +285,18 @@ private void addNewSlave(final AmazonEC2 ec2, final String instanceId) throws Ex
             return; // Wait some more...
 
         final FleetNode slave = new FleetNode(instanceId, "Fleet slave for" + instanceId,
-                fsRoot, "1", Node.Mode.NORMAL, "ec2-fleet", new ArrayList<NodeProperty<?>>(),
+                fsRoot, "1", Node.Mode.NORMAL, this.labelString, new ArrayList<NodeProperty<?>>(),
```
Comment (blmr) [inline src/main/java/com/amazon/jenkins/ec2fleet/EC2FleetCloud.java]: Is there any particular reason why the number of executors is set to "1"?
I'd like to run multiple jobs inside of Docker containers on each slave instance. 
I can only assume that the reason in the logic behind fleet scaling.

Comment (jzila) [inline src/main/java/com/amazon/jenkins/ec2fleet/EC2FleetCloud.java]: I can add a parameter for number of executors.

Comment (blmr) [inline src/main/java/com/amazon/jenkins/ec2fleet/EC2FleetCloud.java]: This feature will be much appreciated ;)

**Comment by jzila** [comment]
@blmr try now (untested, will test tonight)

**Comment by blmr** [comment]
@jzila works like a charm! http://prnt.sc/by7g33
Thanks a lot!

**Comment by jzila** [comment]
It currently won't update the executors on existing nodes, but I can add that later.

**Comment by ryanrupp** [comment]
I'm testing this since I'm running into a lot of the issues mentioned here when trying to use the standard plugin. One thing I ran into with these changes is it's possible for canProvision(label) for the label to be null, see exception stack:

```
Timer task hudson.slaves.NodeProvisioner$NodeProvisionerInvoker@7d4ff5b9 failed
java.lang.NullPointerException
    at com.amazon.jenkins.ec2fleet.EC2FleetCloud.canProvision(EC2FleetCloud.java:400)
    at hudson.slaves.NodeProvisioner$StandardStrategyImpl.apply(NodeProvisioner.java:683)
    at hudson.slaves.NodeProvisioner.update(NodeProvisioner.java:307)
    at hudson.slaves.NodeProvisioner.access$000(NodeProvisioner.java:60)
    at hudson.slaves.NodeProvisioner$NodeProvisionerInvoker.doRun(NodeProvisioner.java:796)
    at hudson.triggers.SafeTimerTask.run(SafeTimerTask.java:50)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
    at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:304)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:178)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)
```

I just added a null check on the label and when doing the logging e.g.:

```
@Override public boolean canProvision(final Label label) { // label can be null
        boolean result = fleet != null && (label == null || Label.parse(this.labelString).containsAll(label.listAtoms()));
        LOGGER.log(Level.FINE, "CanProvision called on fleet: \"" + this.labelString + "\" wanting: \"" + (label != null ?  label.getName() : "(unspecified)") + "\". Returning " + Boolean.toString(result) + ".");
        return result;
    }
```

Thanks for the changes, do you know if these would get merged into the mainline? Not sure if the maintainer is active who is listed on the jenkins plugin.

**Comment by jzila** [comment]
Thanks @ryanrupp.

I'm going to open a new PR from my master, because it looks like this root fork is no longer being maintained.

**Pull request closed without merging**